ext/test/key_provider/key_provider.c

```diff
@@ -0,0 +1,443 @@
+
+    /* Fill buffer by repeatedly copying the pattern */
+    size_t remaining = new_key_size;
+    size_t offset = 0;
+    while (remaining > 0) {
+        const size_t copy_len = (remaining >= pattern_len) ? pattern_len : remaining;
+        memcpy(key_buf + offset, iso_time, copy_len);
+        offset += copy_len;
+        remaining -= copy_len;
+    }
+
+    *new_key = key_buf;
+    *key_size = new_key_size;
+
+    return (0);
+}
+
+/*
+ * kp_rotate_key --
+ *     Rotate the current key by generating a new key with a repeating pattern.
+ */
+static int
+kp_rotate_key(KEY_PROVIDER *kp)
+{
+    /*
+     * Always expiring key is a special case. In this case, we always return the default key to make
+     * sure get_key returns consistent key size. For other expiration policies, we generate a new
+     * key.
+     */
+    if (kp->key_expires != KEY_EXPIRES_ALWAYS) {
+        uint8_t *new_key = NULL;
+        size_t new_key_size = 0;
+        int ret = 0;
+
+        if ((ret = kp_generate_key(&new_key, &new_key_size)) != 0)
+            return (ret);
+
+        /* Free old key and update state */
+        kp_free_key(kp);
+        kp->state.current_key = new_key;
+        kp->state.key_size = new_key_size;
+    }
+
+    kp->state.key_time = clock();
+    kp->state.current_lsn = 0; /* New key, no LSN yet */
+
+    return (0);
+}
+
+/*
+ * kp_get_key --
+ *     Fetch the latest key for checkpoint writes. The WT_CRYPT_KEYS::keys::size should be set to
+ *     zero if the key has not changed. This is called by WiredTiger when creating a checkpoint to
+ *     get the current encryption key.
+ */
+static int
+kp_get_key(WT_KEY_PROVIDER *wtkp, WT_SESSION *session, WT_CRYPT_KEYS *crypt)
+{
+    KEY_PROVIDER *kp = (KEY_PROVIDER *)wtkp;
+    LOG_DEBUG(kp, session, "Current key: LSN=%" PRIu64 ", key_time=%.2f, size=%" PRIzu,
+      kp->state.current_lsn, CLOCK_SECS(kp->state.key_time), kp->state.key_size);
+
+    /*
+     * Real key provider may rotate the key independently of the get_key calls. In the mock
+     * implementation the key is rotated only when its size has been requested. This is to prevent
+     * key rotation between paired get_key calls: first requesting the size, then filling the data.
+     */
+    if (crypt->keys.data == NULL && kp_key_expired(kp)) {
+        LOG_INFO(kp, session, "Key expired (key_time=%.2f)", CLOCK_SECS(kp->state.key_time));
+
+        int ret = kp_rotate_key(kp);
+        if (ret != 0) {
+            LOG_ERROR(kp, session, "Failed to rotate key: %d", ret);
+            return (ret);
+        }
+
+        LOG_INFO(kp, session, "Reporting new key (key_time=%.2f, key_size=%" PRIzu ")",
+          CLOCK_SECS(kp->state.key_time), kp->state.key_size);
+        crypt->keys.size = kp->state.key_size;
+    } else if (crypt->keys.data != NULL) {
+        /* The size of requested data must match previously reported key size. */
+        assert(crypt->keys.size == kp->state.key_size);
+
```

Comment:
With the new idea of WT owning the storage, might there be a race here?
- starting condition, there is 1 KEK key, needing 1000 bytes.
- first call to kp_get_key() -- returns 1000 bytes
- key is rotated, there are 2 KEK keys, needing 2000 bytes.
- second call to kp_get_key() with 1000 byte buffer
- ??

One way to avoid this, is for the second call to only write the first key, and ignore the second for the time being (after all it is a race).  Or, prohibit key rotation during a checkpoint.  Or always extra to be allocated.  Reasonable workarounds, but this should be a known issue for our clients, and maybe a comment here.

Reply:
@ablekhman you mentioned that one solution is that if there is a rotation between the two calls and the size of the struct to be stored is now too big for the buffer, the key provider could return an error.  That's fine, but it's an obligation for the key provider to do this.  (Or to do one of the other solutions I suggest above).  My concern is that obligation is not codified anywhere other than this thread.